### PR TITLE
[WebGPU EP] Add PRelu support

### LINKED ISSUE #29859: [Feature Request] [WebGPU EP] Add PRelu support
### Describe the feature request

**Problem:**
PRelu isn't supported by WebGPU EP.

**Proposal:**
Enable PRelu for WebGPU EP.

**Possible changes needed (actual code pending):**
Need to make sure the versions and types comply with [ONNX PRelu documentation](https://onnx.ai/onnx/operators/onnx__PRelu.html#).

In onnxruntime/core/providers/webgpu/math/binary_elementwise_ops.cc, add the implementation and versioned kernels:
```
WEBGPU_BINARY_IMPL(PRelu, "select(b * a, a, a >= vec4<input_a_element_t>(0))")
WEBGPU_BINARY_KERNEL(PRelu, 6, PRelu, WebGpuSupportedFloatTypes())
WEBGPU_BINARY_VERSIONED_KERNEL(PRelu, 7, 8, PRelu, WebGpuSupportedFloatTypes())
WEBGPU_BINARY_VERSIONED_KERNEL(PRelu, 9, 15, PRelu, WebGpuSupportedNumberTypes())
WEBGPU_BINARY_KERNEL(PRelu, 16, PRelu, WebGpuSupportedNumberTypes())
```

In onnxruntime/core/providers/webgpu/webgpu_execution_provider.cc, add the classes:
```
class ONNX_OPERATOR_KERNEL_CLASS_NAME(kWebGpuExecutionProvider, kOnnxDomain, 6, PRelu);
class ONNX_OPERATOR_VERSIONED_KERNEL_CLASS_NAME(kWebGpuExecutionProvider, kOnnxDomain, 7, 8, PRelu);
class ONNX_OPERATOR_VERSIONED_KERNEL_CLASS_NAME(kWebGpuExecutionProvider, kOnnxDomain, 9, 15, PRelu);
class ONNX_OPERATOR_KERNEL_CLASS_NAME(kWebGpuExecutionProvider, kOnnxDomain, 16, PRelu);
```
Also add the entries to function_table:
```
KERNEL_CREATE_INFO(6, PRelu),
KERNEL_CREATE_INFO_VERSIONED(7, 8, PRelu),
KERNEL_CREATE_INFO_VERSIONED(9, 15, PRelu),
KERNEL_CREATE_INFO(16, PRelu),
```

Need to make sure WebGPU EP PRelu test coverage is OK.

### Describe scenario use case

Models with PRelu and only GPU assigned subgraphs might be able to be run with graph capture enabled.

### PR DESCRIPTION
**Pull request by AngelGalindo7**
### Description

Add PRelu support for the WebGPU EP: kernel implementation plus registrations for
opsets 7-8, 9-15 and 16. Opset 6 is not registered, matching the CPU and CUDA EPs.

Coverage: the existing ActivationOpTest.PRelu* cases fan out to every registered EP,
so they exercise the new kernel on the scalar/no-broadcast and general-broadcast paths
in BinaryElementwise, plus the inf/NaN edge cases. Those run at OpTester's default
opset (7), so two new float16 cases cover the 9-15 and 16 registrations; the opset-9
case uses a last dim of 4, which is also the only coverage of the vectorized path.
Both use float16 because the CPU EP's PRelu is float-only, so they cannot silently
pass via CPU fallback.

### Motivation and Context

Fixes #29859

Without it, PRelu falls back to the CPU EP and Memcpy nodes get inserted around it
when it sits between WebGPU nodes. That costs a GPU→CPU→GPU round trip per inference,
and causes a session-init failure when graph capture is enabled — the scenario named
in the issue.

### CODE REVIEW

**Comment by AngelGalindo7** [comment]
@microsoft-github-policy-service agree

**Comment by AngelGalindo7** [comment]
@hariharans29 can you please take a look and help trigger the CI workflows? Thanks!

File: onnxruntime/core/providers/webgpu/math/binary_elementwise_ops.cc [hunk 1]
```diff
@@ -438,6 +438,11 @@ WEBGPU_BINARY_VERSIONED_KERNEL_2(Pow, 12, 12, Pow, WebGpuSupportedNumberTypes(),
 WEBGPU_BINARY_VERSIONED_KERNEL_2(Pow, 13, 14, Pow, WebGpuSupportedNumberTypes(), WebGpuSupportedNumberTypes())
 WEBGPU_BINARY_KERNEL_2(Pow, 15, Pow, WebGpuSupportedNumberTypes(), WebGpuSupportedNumberTypes())
 
+WEBGPU_BINARY_IMPL(PRelu, "select(b * a, a, a >= vec4<input_a_element_t>(0))")
+WEBGPU_BINARY_VERSIONED_KERNEL(PRelu, 7, 8, PRelu, WebGpuSupportedFloatTypes())
+WEBGPU_BINARY_VERSIONED_KERNEL(PRelu, 9, 15, PRelu, WebGpuSupportedNumberTypes())
+WEBGPU_BINARY_KERNEL(PRelu, 16, PRelu, WebGpuSupportedNumberTypes())
```
[the remark this answers is not in the document]
Comment (AngelGalindo7) [reply onnxruntime/core/providers/webgpu/math/binary_elementwise_ops.cc]: Narrowed both to WebGpuSupportedFloatTypes(). Flagging that this deviates from #29859, which specifies WebGpuSupportedNumberTypes() for opsets 9-15 and 16.

On the schema point: the float-only constraint applies to opset 7-8 only. Opset 9 widened T to tensor(float16), tensor(float), tensor(double), tensor(uint32), tensor(uint64), tensor(int32), tensor(int64), and the constraint description changed from "Constrain input and output types to float tensors" to "float/int tensors". PRelu-16 adds bfloat16 on top. Per-version constraints are rendered at [onnx.ai/onnx/operators/onnx__PRelu.html](https://onnx.ai/onnx/operators/onnx__PRelu.html).

So the original 7-8 vs 9+ split was tracking that opset 9 change, per the issue. Note too that WebGpuSupportedFloatTypes() does not match the schema either, since every PRelu opset including 7 permits tensor(double) and WGSL has no f64. No registration here matches the schema exactly.

I have narrowed anyway since [onnxruntime/core/providers/cpu/math/element_wise_ops.h](https://github.com/microsoft/onnxruntime/blob/18e3e86df23b75807b773a7cc993310457790cd8/onnxruntime/core/providers/cpu/math/element_wise_ops.h#L487-L490) records a decision to keep PRelu float-only:

```
// Currently only registered for float. Integer types would require addressing
// overflow in multiplication and avoiding 0.0*inf=NaN in branchless formulations.
static_assert(std::is_floating_point_v<T>,
              "PRelu is only supported for floating-point types.");
```

It also makes the registered surface and the tests line up exactly. Every registration is now exercised: opset 7-8 by the existing ActivationOpTest.PRelu* cases plus the newly enabled test_prelu_example and test_prelu_broadcast node tests, opset 9-15 by PRelu_Float16_Opset9_WebGpu, and opset 16 by PRelu_Float16_Opset16_WebGpu. Keeping WebGpuSupportedNumberTypes() would leave int32 and uint32 with no coverage anywhere, since CPU PRelu is float-only and cannot back a shared test.

Happy to open a follow up issue to restore the integer types with int32 and uint32 coverage, which would also need a decision on overflow semantics for the multiply. Happy to revert this narrowing if you would rather the PR match #29859 as written.

**Comment by AngelGalindo7** [comment]
@hariharans29 would you be able to take a look, or help trigger the CI pipelines? Only the CLA check has run so far. Thanks!

**Comment by hariharans29** [comment]
> @hariharans29 would you be able to take a look, or help trigger the CI pipelines? Only the CLA check has run so far. Thanks!

Appreciate your contibution. Please allow us some time to review PRs as some of us are juggling multiple areas. Do give us a shout if something takes more than a few days. As for the CIs, they did run the first time

image

They haven't run the second time (after your recent commits). I will trigger them now.

File: onnxruntime/core/providers/webgpu/math/binary_elementwise_ops.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (kareertl) [reply onnxruntime/core/providers/webgpu/math/binary_elementwise_ops.cc]: I'm personally OK with the narrow set of types, but it's preferrable to support more types. The AI comment was interesting. Good job on checking the schema.

**Comment by hariharans29** [comment]
# Review: PR #30512 — [WebGPU EP] Add PRelu support (head `3b5cfff`)

Author: `@AngelGalindo7` (external contributor, CLA signed). 2 commits, `+70 / -2` across 4 files. **CI: 86 / 86 checks OK** on both commits. **Closes issue #29859** ("Feature Request: WebGPU PRelu support"). Reviewers: Copilot AI (1 comment, resolved by narrowing types), `@kareertl` (informal thumbs-up), `@edgchen1` (added `ep:WebGPU` label), `@hariharans29` (engaged, triggered second CI run).

**Verdict: approve.** Textbook contribution. Correct WGSL semantics, uses the existing `BinaryElementwise` infrastructure to inherit broadcasting, type-constraint narrowing was thoughtfully argued through and lands in the right place, and tests are designed to positively assert the WebGPU kernel path (not silently pass on CPU fallback). Also resolves the real bug that motivated the issue: PRelu falling back to CPU inserts Memcpy nodes that break WebGPU graph capture.

---

## What this PR does

Four files, minimal surface:

1. **[onnxruntime/core/providers/webgpu/math/binary_elementwise_ops.cc](onnxruntime/core/providers/webgpu/math/binary_elementwise_ops.cc)** — the PRelu implementation:
   ```cpp
   WEBGPU_BINARY_IMPL(PRelu, "select(b * a, a, a >= vec4<input_a_element_t>(0))")
   WEBGPU_BINARY_VERSIONED_KERNEL(PRelu, 7, 8,  PRelu, WebGpuSupportedFloatTypes())
   WEBGPU_BINARY_VERSIONED_KERNEL(PRelu, 9, 15, PRelu, WebGpuSupportedFloatTypes())
   WEBGPU_BINARY_KERNEL(PRelu, 16, PRelu, WebGpuSupportedFloatTypes())
   ```

2. **[onnxruntime/core/providers/webgpu/webgpu_execution_provider.cc](onnxruntime/core/providers/webgpu/webgpu_execution_provider.cc)** — three `KERNEL_CREATE_INFO` entries for opsets 7-8, 9-15, and 16.

3. **[js/web/test/suite-test-list.jsonc](js/web/test/suite-test-list.jsonc)** — un-comments `test_prelu_broadcast` and `test_prelu_example` (ONNX backend node tests).

4. **[onnxruntime/test/providers/cpu/activation/activation_op_test.cc](onnxruntime/test/providers/cpu/activation/activation_op_test.cc)** — two new WebGPU-only fp16 tests: `PRelu_Float16_Opset9_WebGpu` and `PRelu_Float16_Opset16_WebGpu`.

---

## Correctness — WGSL expression matches ONNX PRelu

ONNX PRelu: $y = x \cdot [x \geq 0] + \alpha \cdot x \cdot [x < 0] = \begin{cases} x & \text{if } x \geq 0 \\ \alpha \cdot x & \text{if } x < 0 \end{cases}$

The WGSL expression:
```wgsl
select(b * a, a, a >= vec4<input_a_element_t>(0))
```

WGSL `select(false_val, true_val, cond)` semantics:
- Where `a >= 0` (per lane): returns `a` — this is `x` untouched.
- Where `a < 0`  (per lane): returns `b * a` — this is `slope * x`.

Mapping: `a` = X (input), `b` = slope. So per lane:
- `x >= 0` → `y = x`. ✓
- `x <  0` → `y = slope * x`. ✓

Component-wise: `vec4<input_a_element_t>(0)` constructs a 4-lane zero vector of the input's element type. Comparison `a >= vec4<0>` returns `vec4<bool>`, and `select` with a `vec4<bool>` condition does per-lane selection. Correct vectorization pattern matching the surrounding `WEBGPU_BINARY_IMPL` conventions (see `Pow`, `Equal`, etc.).

### Edge cases

- **`x = -inf, slope = 0`** (existing `PRelu_NegInf_ZeroSlope` test): `b * a = 0 * -inf = NaN` per IEEE 754. `a >= 0` is false. `select` returns `NaN`. This matches CPU behavior and the intent of the existing test that asserts the NaN result.
- **`x = 0`**: `a >= 0` is true (WGSL comparison includes equality) → returns `0`. Matches ONNX spec (0 counts as non-negative).
- **`x = NaN`**: `NaN >= 0` is false in WGSL → returns `slope * NaN = NaN`. Correct propagation.

### Broadcasting inherited from the framework

`WEBGPU_BINARY_IMPL` folds the expression into `BinaryElementwise`, which already handles NumPy-style broadcasting. PRelu's slope shape (typically `[C]` or `[C, 1, 1]` for NCHW inputs) broadcasts to X via the same infrastructure that handles `Add`, `Mul`, etc. — no PRelu-specific broadcast code needed. Clean design leverage. ✓

---

## Type constraint narrowing — well-argued exchange with Copilot AI

Initial PR at `1c06897` used `WebGpuSupportedNumberTypes()` for opsets 9-15 and 16, matching the request in issue #29859. Copilot AI flagged this as overly permissive: ONNX PRelu opset 7-8 is float-only, and while opsets 9+ widen the schema to include integer types, the WebGPU kernel doesn't handle integer overflow/NaN semantics correctly.

The author's response was excellent — cited the CPU kernel's explicit design decision from [onnxruntime/core/providers/cpu/math/element_wise_ops.h](onnxruntime/core/providers/cpu/math/element_wise_ops.h):

```cpp
// Currently only registered for float. Integer types would require addressing
// overflow in multiplication and avoiding 0.0*inf=NaN in branchless formulations.
static_assert(std::is_floating_point_v<T>,
              "PRelu is only supported for floating-point types.");
```

...and adopted the narrower constraint (`WebGpuSupportedFloatTypes()` = {float, fp16}) at head `3b5cfff` ("Narrow PRelu opset 9+ registrations to float types"). Author also flagged that `WebGpuSupportedFloatTypes()` still doesn't exactly match the ONNX schema (which includes double, but WGSL has no f64) and offered to file a follow-up issue for integer support.

This is model contributor behavior. `@kareertl` remarked: *"I'm personally OK with the narrow set of types, but it's preferrable to support more types. The AI comment was interesting. Good job on checking the schema."*

The narrower constraint also has a nice test-coverage property: **every registration is now positively exercised** (see below), whereas `WebGpuSupportedNumberTypes()` would have left int32/uint32 with no coverage since the CPU EP can't back a shared test.

---

## Test design — positively asserts WebGPU across all three opset ranges

Coverage matrix:

| Registration | Test | Type | Path exercised |
|---|---|---|---|
| Opset 7-8 | `ActivationOpTest.PRelu*` (existing) | float32 | Scalar broadcast + inf/NaN edges |
| Opset 7-8 | `test_prelu_example`, `test_prelu_broadcast` (ONNX backend) | float32 | General broadcast |
| Opset 9-15 | `PRelu_Float16_Opset9_WebGpu` | **float16** | Per-channel broadcast, last dim = 4 (vectorized) |
| Opset 16 | `PRelu_Float16_Opset16_WebGpu` | **float16** | Per-channel broadcast, last dim = 3 (tail-vec4 path) |

**Critical design choice**: the new tests use **fp16** because the CPU EP's PRelu is float32-only. This means the test cannot silently pass on CPU fallback — the WebGPU kernel *must* run. This is exactly the positive-assert pattern I've been recommending in reviews of similar WebGPU PRs.

Test bodies use `execution_providers.push_back(std::move(webgpu_ep))` + `test.Run(..., &execution_providers)` — the standard WebGPU-only test dispatch. Tolerance `SetOutputAbsErr("Y", 0.01f)` + `SetOutputRelErr("Y", 0.01f)` = ~10 fp16 ULP, generous but appropriate.

Both tests use per-channel slope broadcast (shape `[2, 1, 1]` broadcast to `[1, 2, H, W]` input), which is the canonical CNN use case. `PRelu_Float16_Opset9_WebGpu` includes a **negative** slope value (`-0.25`) — technically legal per the ONNX schema and exercises the sign of the multiply. `PRelu_Float16_Opset16_WebGpu` uses last-dim = 3 to exercise the non-multiple-of-4 tail path.

Author's PR description explicitly reasons through this: *"Both use float16 because the CPU EP's PRelu is float-only, so they cannot silently pass via CPU fallback."*

---

## Impact — closes a real bug, not just a feature gap

The linked issue #29859 notes that without a WebGPU PRelu kernel:
1. PRelu falls back to CPU EP.
2. Memcpy nodes get inserted around the CPU PRelu when it sits between WebGPU nodes.
3. Each inference incurs a GPU→CPU→GPU round trip.
4. **Session init fails when graph capture is enabled** (graph capture can't handle Memcpy insertions).

Point 4 is the critical scenario — WebGPU graph capture is a significant performance feature and this PR unblocks models using PRelu (SegFormer, several YOLO variants, some pose-estimation networks) from using it.

---

## Nits (all non-blocking)

1. **`vec4<input_a_element_t>(0)` — integer-to-float literal conversion**. WGSL allows this for a concrete scalar type with an abstract-int literal, but `vec4<f32>(0.0)` is marginally clearer. Cosmetic.
2. **Second CI trigger required manual push from `@hariharans29`.** First-time contributors' PRs don't auto-rerun CI on subsequent pushes; `@hariharans29` handled this correctly and both commits are 86/86 OK. Not a PR issue.
3. **Follow-up integer type support**. Author offered to file a follow-up issue for int32/uint32 support with proper overflow/NaN semantics. Good to accept that offer — the CPU kernel's `static_assert` and comment give a clear template for what such a PR would need to handle.
4. **JS suite entries**. Enabling `test_prelu_broadcast` and `test_prelu_example` is exactly right — these are the ONNX Node Tests that exercise the JS runtime path end-to-end.

---

## Merge state

- Both commits **86 / 86 checks OK**.
- Copilot AI comment resolved (author narrowed types).
- `@kareertl` gave informal thumbs-up.
- `@edgchen1` tagged with `ep:WebGPU` label.
- `@hariharans29` engaged, triggered second CI run, has not yet formally approved.
- Awaits Microsoft-side approving review.

## Bottom line

Small, focused, correct, well-tested, well-argued. Uses existing infrastructure (`BinaryElementwise`) rather than duplicating machinery. Type-constraint decision has substance behind it (cites CPU kernel's design rationale). Tests positively assert WebGPU kernel selection via fp16 (CPU EP can't back fp16 PRelu). Closes a long-standing issue #29859 and unblocks graph-capture inference for models with PRelu.

**Action items:**
1. `@hariharans29` or `@edgchen1` — approve on green CI.
2. Optional post-merge: author files follow-up issue for integer PRelu support, addressing overflow-in-multiplication and `0.0 * inf = NaN` in a branchless formulation (using the CPU comment as a template).

Ready to merge.

**Comment by hariharans29** [comment]
@AngelGalindo7 - Thanks for waiting - The PR LGTM. :) There are some non-blocking nits listed above. Please see if any of that applies and submit them as a future PR or along with another future PR. Thanks for the contribution - merging this.

**Review by hariharans29 — APPROVED** [review]
